tutorial: Deploying Lambda container image locally with ECR using LocalStack (#351)
localstack/docs: content/en/tutorials/lambda-ecr-container-images/index.md

- [awslocal](https://docs.localstack.cloud/integrations/aws-cli/#localstack-aws-cli-awslocal)
- [Docker](https://docker.io/)

## Creating a Lambda function

To package & deploy a Lambda as a container image, we must need to create a Lambda function containing our code and a Dockerfile. Create a new directory and initialize create two files: `handler.py`, to save our Python-based Lambda function, and `Dockerfile`, to package our code and dependencies into a container image.

{{< command >}}
$ mkdir -p lambda-container-image
$ cd lambda-container-image
$ touch handler.py Dockerfile
{{< / command >}}

Let us use the following Python code to create a Lambda function that returns a simple `'Hello from LocalStack Lambda container image!'` message.

```python
def handler(event, context):
    print('Hello from LocalStack Lambda container image!')
```

In the above example, the `handler` function is executed by the Lambda service every time a trigger event occurs. The above function serves as an entrypoint for the Lambda function inside a runtime environment and accepts `event` and `context`, to receive information about the event and the invocation properties.

## Building the image

> To package our Lambda function as a container image, we need to create a Dockerfile that contains the instructions to build our image. The image should be able to execute a read-only file system with access to a `/tmp` directory. We would be using [AWS base images for Lambda](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-images.html#runtimes-images-lp) to preload the runtimes and dependencies that are required to create Lambda images.

Reviewer: ```suggestion
To package our Lambda function as a container image, we need to create a `Dockerfile` that contains the instructions to build our image. The image should be able to execute a read-only file system with access to a `/tmp` directory. We would be using [AWS base images for Lambda](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-images.html#runtimes-images-lp) to preload the runtimes and dependencies that are required to create Lambda images.
```

To build the container image for Python, we would use a `python:3.8 base image`. Open the `Dockerfile` and specify the function handler to ensure that the Lambda runtime can locate it where the Lambda handler is available:

```Dockerfile
FROM public.ecr.aws/lambda/python:3.8

COPY ./handler.py ./

CMD [ "handler.handler" ]
```

{{< alert title="Note" color="primary">}}
If you have additional dependencies specificed, create a file named `requirements.txt` and list the libraries in the file. Install the dependencies in the Dockerfile under the `${LAMBDA_TASK_ROOT}` directory.
{{< /alert >}}

Use Docker to build an image using this function code:

{{< command >}}
$ docker build -t localstack-lambda-container-image .
{{< / command >}}

## Publishing the image to ECR

> Now that the initial setup is done, we can give LocalStack's AWS emulation a try to push our image to ECR and deploy the Lambda container image. Let's start LocalStack:

Reviewer: ```suggestion
Now that the initial setup is done, we can give LocalStack's AWS emulation a try by pushing our image to ECR and deploying the Lambda container image. Let's start LocalStack:
```

{{< command >}}

> $ LOCALSTACK_API_KEY=<your-api-key> localstack start -d

Reviewer: nit: not sure this is really required - I think we can generally assume that LocalStack is already started (or refer to the getting started section). Thoughts?

Author: I believe that we can still keep this section around, since the tutorial serve the information end-to-end and in this way, since we can expect people to try out these tutorials without even looking at rest of the documentation. What do you think?

To deploy the container image as a Lambda function, we need to create a new Lambda function using the `create-function` command. We will pass the `ImageUri` flag to specify the image URI of the container image that we have pushed to the ECR repository.

{{< command >}}
$ awslocal lambda create-function \
    --function-name localstack-lambda-container-image \
    --package-type Image \
    --code ImageUri="localstack-lambda-container-image" \
    --role arn:aws:iam::000000000:role/lambda-r1 \
    --handler handler.handler
{
    "FunctionName": "localstack-lambda-container-image",
    "FunctionArn": "arn:aws:lambda:<REGION>:000000000000:function:localstack-lambda-container-image",
    "Role": "arn:aws:iam::000000000:role/lambda-r1",
    "Handler": "handler.handler",
    "Description": "",
    "Timeout": 3,
    "LastModified": "<TIMESTAMP>",
    "Version": "$LATEST",
    "VpcConfig": {},
    "TracingConfig": {
        "Mode": "PassThrough"
    },
    "RevisionId": "<REVISION_ID>",
    "Layers": [],
    "State": "Active",
    "LastUpdateStatus": "Successful",
    "PackageType": "Image",
    "Architectures": [
        "x86_64"
    ]
}
{{< / command >}}

> The above command has taken various flags to create the Lambda function. We have specified the `ImageUri` flag to specify the image URI of the container image that we have pushed to the ECR repository. We have also specified the `package-type` flag to specify the package type as `Image`. For the role, we have specified a mock role ARN. To create an actual role, refer to the [IAM documentation]({{< ref "iam" >}}).

Reviewer: ```suggestion
The above command has taken various flags to create the Lambda function. We have specified the `ImageUri` flag to specify the image URI of the container image that we have pushed to the ECR repository. We have also specified the `package-type` flag to specify the package type as `Image`. For the role, we have specified a mock role ARN. To create an actual role, please refer to the [IAM documentation]({{< ref "iam" >}}).
```

> Let us know invoke the Lambda function using the `invoke` command:

Reviewer: ```suggestion
Let us now invoke the Lambda function using the `invoke` command:
```

{{< command >}}
$ awslocal lambda invoke --function-name localstack-lambda-container-image /tmp/lambda.out
{
    "StatusCode": 200,
    "LogResult": "",
    "ExecutedVersion": "$LATEST"
}
{{< / command >}}

The logs of the Lambda invocation should be visible in the LocalStack container output (with `DEBUG=1` enabled):

{{< command >}}
Starting XRay server loop on UDP port 2000
Starting DNS server loop on UDP port 53
-----
Hello from LocalStack Lambda container image!
{{< / command >}}

## Conclusion

> With the Lambda container image support, you can use Docker to package your custom code and dependencies for Lambda functions. LocalStack allows you to package, deploy, and invoke Lambda functions locally. Using LocalStack, you can develop, debug, and test your Lambda functions in conjunction with a wide range of AWS services. Check out [Lambda Hot Swapping]({{< ref "hot-swapping" >}}) and [Lambda Hot Reloading]({{< ref "debugging" >}})

Reviewer: ```suggestion
With the Lambda container image support, you can use Docker to package your custom code and dependencies for Lambda functions. LocalStack allows you to package, deploy, and invoke Lambda functions locally. Using LocalStack, you can develop, debug, and test your Lambda functions in conjunction with a wide range of AWS services. Check out [Lambda Hot Swapping]({{< ref "hot-swapping" >}}) and [Lambda Hot Reloading]({{< ref "debugging" >}}) for advanced usage patterns.
```
